File: LList.cpp
#include <iostream>
#include <string>
#include "LList.h"
using namespace std;

void LCREATE(LList& new_LList, const string& data){
    new_LList.head = new LLNode(data, nullptr, nullptr);
    new_LList.tail = new_LList.head;
}

LLNode* LGETINDEX(LList& list, int index){
    LLNode* Address = list.head;
    while(1){
        if (index == 0){
            return Address;
        }
        if(Address -> next == nullptr){
            if (index != 0){
                cout << "Элемента с таким индексом нет" << endl;
                exit(0);
            }
            return Address;
        }
        else{
            Address = Address -> next;
            index--;
        }
    }
}

LLNode* LGETDATA(LList& list, const string& data){
    LLNode* Address = list.head;
    while(1){
        if (Address -> data == data){
            return Address;
        }
        if(Address -> next == nullptr){
            if (Address -> data != data){
                cout << "Узел отсутствует в списке" << endl;
                exit(0);
            }
            return Address;
        }
        else{
            Address = Address -> next;
        }
    }
}


void LADDTAIL(LList& list, const string& data){
    LLNode* new_LLNode = new LLNode(data, list.tail, nullptr);
    list.tail -> next = new_LLNode;
    list.tail = new_LLNode;
}


void LADDHEAD(LList& list, const string& data){
    LLNode* new_LLNode = new LLNode(data, nullptr, list.head);
    
    // Обновляем previous у старой головы
    if (list.head != nullptr) {
        list.head->previous = new_LLNode;
    }
    
    list.head = new_LLNode;
    
    // Если список был пуст (tail = nullptr), обновляем tail
    if (list.tail == nullptr) {
        list.tail = new_LLNode;
    }
}

void LADDAFTER(LList& list, LLNode* ptr, const string& data){
    if (ptr -> next == nullptr){
        LADDTAIL(list, data);
        return;
    }
    LLNode* new_LLNode = new LLNode(data, ptr, ptr -> next);
    ptr -> next = new_LLNode;
    new_LLNode -> next -> previous = new_LLNode;
}

void LADDBEFORE(LList& list, LLNode* ptr, const string& data){
    if (list.head == ptr){
        LADDHEAD(list, data);
        return;
    }

    LLNode* new_LLNode = new LLNode(data, ptr->previous, ptr);
    
    // Обновляем указатель next у предыдущего узла
    ptr->previous->next = new_LLNode;
    
    // Обновляем указатель previous у текущего узла
    ptr->previous = new_LLNode;
}

//--------------------------------------

void LDELTAIL(LList& list){
    LLNode* Address = list.tail -> previous;
    delete list.tail;
    list.tail = Address;
    Address -> next = nullptr;
}


void LDELHEAD(LList& list){
    LLNode* Address = list.head -> next;
    delete list.head;
    list.head = Address;
}

void LDELAFTER(LList& list, LLNode* ptr){
    if (ptr -> next == nullptr) return;
    if (ptr -> next == list.tail) {
        LDELTAIL(list);
        return;
    }
    LLNode* Address = ptr -> next -> next;
    delete ptr -> next;
    ptr -> next = Address;
    Address -> previous = ptr;
}

void LDELBEFORE(LList& list, LLNode* ptr){
    if (list.head == ptr){
        cout << "Элемент первый в списке" << endl;
        return;
    }
    if (ptr -> previous == list.head) {
        LDELHEAD(list);
        return;
    }
    LLNode* Address = ptr -> previous -> previous;
    delete ptr -> previous;
    ptr -> previous = Address;
    Address -> next = ptr;
}

void LDELDATA(LList& list, const string& data){
    LLNode* Address = list.head;
    if (Address -> data == data){
        LDELHEAD(list);
        return;
    }

    while(1){

        if(Address -> next == nullptr){
            if (Address -> data == data){
                LDELTAIL(list);
                return;
            }
            cout << "Элемента нет в списке" << endl;
            return;
        }

        if (Address -> data == data){
            Address -> previous -> next = Address -> next;
            Address -> next -> previous = Address -> previous;
            delete Address;
            return;
        }

        Address = Address -> next;
    }
}

void LPRINT(LList& list){
    LLNode* Address = list.head;
    while (1){
        if(Address -> next == nullptr){
            cout << Address -> data << endl;
            return;
        }
        else{
            cout << Address -> data << endl;
            Address = Address -> next;
        }
    }
}

File: LList.go
package main

import (
	"fmt"
)

type LLNode struct {
	data     string
	previous *LLNode
	next     *LLNode
}

// Конструктор
func NewLLNode(val string, prev *LLNode, nxt *LLNode) *LLNode {
	return &LLNode{
		data:     val,
		previous: prev,
		next:     nxt,
	}
}

type LList struct {
	tail *LLNode
	head *LLNode
}

// Конструктор
func NewLList() *LList {
	return &LList{
		tail: nil,
		head: nil,
	}
}

// Конструктор копирования
func (l *LList) Copy() *LList {
	newList := &LList{
		tail: nil,
		head: nil,
	}

	if l.head != nil {
		currentOther := l.head
		var last *LLNode

		// Копируем первый элемент
		newList.head = NewLLNode(currentOther.data, nil, nil)
		last = newList.head
		currentOther = currentOther.next

		// Копируем остальные элементы
		for currentOther != nil {
			newNode := NewLLNode(currentOther.data, last, nil)
			last.next = newNode
			last = newNode
			currentOther = currentOther.next
		}
		newList.tail = last
	}

	return newList
}

// Оператор присваивания
func (l *LList) Assign(other *LList) *LList {
	if l != other {
		// Очищаем текущий список
		current := l.head
		for current != nil {
			next := current.next
			current = nil
			current = next
		}
		l.head = nil
		l.tail = nil

		// Копируем элементы из другого списка
		if other.head != nil {
			currentOther := other.head
			var last *LLNode

			l.head = NewLLNode(currentOther.data, nil, nil)
			last = l.head
			currentOther = currentOther.next

			for currentOther != nil {
				newNode := NewLLNode(currentOther.data, last, nil)
				last.next = newNode
				last = newNode
				currentOther = currentOther.next
			}
			l.tail = last
		}
	}
	return l
}

func LCREATE(newLList *LList, data string) {
	newLList.head = NewLLNode(data, nil, nil)
	newLList.tail = newLList.head
}

func LGETINDEX(list *LList, index int) *LLNode {
	Address := list.head
	for {
		if index == 0 {
			return Address
		}
		if Address.next == nil {
			if index != 0 {
				fmt.Println("Элемента с таким индексом нет")
				return nil
			}
			return Address
		} else {
			Address = Address.next
			index--
		}
	}
}

func LGETDATA(list *LList, data string) *LLNode {
	Address := list.head
	for {
		if Address.data == data {
			return Address
		}
		if Address.next == nil {
			if Address.data != data {
				fmt.Println("Узел отсутствует в списке")
				return nil
			}
			return Address
		} else {
			Address = Address.next
		}
	}
}

func LADDTAIL(list *LList, data string) {
	newLLNode := NewLLNode(data, list.tail, nil)
	list.tail.next = newLLNode
	list.tail = newLLNode
}

func LADDHEAD(list *LList, data string) {
	newLLNode := NewLLNode(data, nil, list.head)

	// Обновляем previous у старой головы
	if list.head != nil {
		list.head.previous = newLLNode
	}

	list.head = newLLNode

	// Если список был пуст (tail = nil), обновляем tail
	if list.tail == nil {
		list.tail = newLLNode
	}
}

func LADDAFTER(list *LList, ptr *LLNode, data string) {
	if ptr.next == nil {
		LADDTAIL(list, data)
		return
	}
	newLLNode := NewLLNode(data, ptr, ptr.next)
	ptr.next = newLLNode
	newLLNode.next.previous = newLLNode
}

func LADDBEFORE(list *LList, ptr *LLNode, data string) {
	if list.head == ptr {
		LADDHEAD(list, data)
		return
	}

	newLLNode := NewLLNode(data, ptr.previous, ptr)

	// Обновляем указатель next у предыдущего узла
	ptr.previous.next = newLLNode

	// Обновляем указатель previous у текущего узла
	ptr.previous = newLLNode
}

func LDELTAIL(list *LList) {
	if list.tail == nil {
		return
	}
	Address := list.tail.previous
	list.tail = Address
	if Address != nil {
		Address.next = nil
	} else {
		// Если удалили последний элемент
		list.head = nil
	}
}

func LDELHEAD(list *LList) {
	if list.head == nil {
		return
	}
	Address := list.head.next
	list.head = Address
	if Address != nil {
		Address.previous = nil
	} else {
		// Если удалили последний элемент
		list.tail = nil
	}
}

func LDELAFTER(list *LList, ptr *LLNode) {
	if ptr.next == nil {
		return
	}
	if ptr.next == list.tail {
		LDELTAIL(list)
		return
	}
	Address := ptr.next.next
	ptr.next = Address
	if Address != nil {
		Address.previous = ptr
	}
}

func LDELBEFORE(list *LList, ptr *LLNode) {
	if list.head == ptr {
		fmt.Println("Элемент первый в списке")
		return
	}
	if ptr.previous == list.head {
		LDELHEAD(list)
		return
	}
	Address := ptr.previous.previous
	ptr.previous = Address
	if Address != nil {
		Address.next = ptr
	}
}

func LDELDATA(list *LList, data string) {
	if list.head == nil {
		fmt.Println("Список пуст")
		return
	}

	Address := list.head
	if Address.data == data {
		LDELHEAD(list)
		return
	}

	for {
		if Address.next == nil {
			if Address.data == data {
				LDELTAIL(list)
				return
			}
			fmt.Println("Элемента нет в списке")
			return
		}

		if Address.data == data {
			Address.previous.next = Address.next
			Address.next.previous = Address.previous
			return
		}

		Address = Address.next
	}
}

func LPRINT(list *LList) {
	Address := list.head
	for {
		if Address.next == nil {
			fmt.Println(Address.data)
			return
		} else {
			fmt.Println(Address.data)
			Address = Address.next
		}
	}
}
